---
title: "FBA Sensitivity Analysis prototrophic and auxotrophic strains of E. coli"
author: "A. Sorokin"
date:  '`r format(Sys.time(), "%d.%m.%Y")`'
output:
  pdf_document:
    keep_tex: yes
    number_sections: yes
    toc: yes
    toc_depth: 4
    extra_dependencies: ["flafter"]
  word_document: default
  html_document: default
bibliography: FBA-PRCC-EcoliK12.bib
csl: biomed-central.csl
header-includes:
- \usepackage[utf8]{inputenc}
- \usepackage[english]{babel}
- \usepackage{grffile}
- \usepackage{rotating}
- \usepackage{caption}
- \usepackage{longtable}
- \usepackage{lscape}
- \DeclareCaptionLabelFormat{addS}{#1 S#2}
- \captionsetup[table]{name=Supplementary Table}
- \renewcommand{\thetable}{S\arabic{table}}
- \captionsetup[figure]{name=Supplementary Figure}
- \renewcommand{\thefigure}{S\arabic{figure}}
---
```{r loadPackages, include=FALSE, cache=FALSE}
## load additional packages in this chunk
library(pander)
library(knitr)
library(data.table)
library(ggplot2)
library(xtable)
library(tidyr)
library(dplyr)
library(openxlsx)
library(sensitivity)
library(UpSetR)
library(ODEsensitivity)
library(RColorBrewer)
```

```{r setup, include=FALSE, cache=FALSE}
## This chunk should contain global configuration commands.
## Use this to set knitr options and related things. Everything
## in this chunk will be included in an appendix to document the
## configuration used.
#output <- opts_knit$get("rmarkdown.pandoc.to")
knitr::opts_knit$set(stop_on_error = 2L)

## By default R code is only included in HTML versions of the report
## (where it can be collapsed). You can generate a PDF version
## using rmarkdown::pdf_document to get a copy for print. Extensive
## chunks of R code may or may not be desired in /hat setting. If you
## want them simply change the following arguments to `echo = TRUE`.
## In either case the default can be overwritten for individual chunks.
#opts_chunk$set(echo = output=="html")
#opts_chunk$set(warning = output=="html")
#opts_chunk$set(message = output=="html")

## Cache options
knitr::opts_chunk$set(cache=TRUE,warning=FALSE)

## Set 'hide.fig.code' to FALSE to include code chunks that
## produce Figures in the output. Note that this affects all chunks
## that provide a figure caption.
knitr::opts_chunk$set(hold=TRUE, hide.fig.code=FALSE)

## Pander options
pander::panderOptions("digits", 3)
pander::panderOptions("table.split.table", 160)
```

```{r functions, include=FALSE,cache=FALSE}
## Custom functions used in the analysis should go into this chunk.
## They will be listed in their own section of the appendix.

##==================== Functions ====================##
getPRCC<-function(file){
    out<-readRDS(file)
    idx<-which(sapply(out,nrow)>0)
    #l<-lapply(out[idx],function(.x).x[which(rownames(.x)=='Solution'),])
    res<-do.call(rbind,out[idx])
    return(res)
}

normPRCC<-function(prcc){
    prime <- 0.5*log(abs((1+prcc)/(1-prcc)))
    return(prime)
}
signifPRCC<-function(prcc1,prcc2,N1,N2){
    p1<-length(prcc1)-1
    p2<-length(prcc2)-1
    
    z<-(normPRCC(prcc1)-normPRCC(prcc2))/sqrt(1/(N1-3-p1) + 1/(N2-3-p2))
}

printTable <-
  function(mat,main,landscape = TRUE,digits = 0,sig = 0.01,align= 'lllrr',
           label=paste0('tab:',substr(sub(' ','_',main),0,10)),
           include.rownames=FALSE) {
    addtorow          <- list()
    addtorow$pos      <- list()
    addtorow$pos[[1]] <- c(0)
    addtorow$command  <- c(
      paste(
        "\\hline \n",
        "\\endhead \n",
        "\\hline \n",
        "\\multicolumn{3}{l}{\\footnotesize Continued on next page} \n",
        "\\endfoot \n",
        "\\endlastfoot \n",sep = ""
      )
    )
    if (landscape) {
      cat(
        sprintf(
          "\\newpage\n  \\begin{landscape} \n\\begin{center}\n\\captionof{table}{%s (%d)}\n\\scriptsize",
          main,dim(mat)[1]
        )
      )
    }else{
      cat(
        sprintf(
          "\\begin{center}\n\\captionof{table}{%s (%d)}\n\\scriptsize",
          main,dim(mat)[1]
        )
      )
    }
    #cat(dim(mat),names(mat),'\n')
    matU<-mat
    # matU$name<-sanitizestr(mat$name)
    # matU$description<-sanitizestr(mat$description)
    # cat(dim(matU),names(matU),'\n')
    print(
      xtable(
        matU,
        align = align,#paste(align,collapse = ''),
        digits = digits,auto = TRUE)#,label = label)
      ,size = "small",include.colnames = TRUE,
      tabular.environment = "longtable", sanitize.text.function=NULL,#function(.x)sanitizestr(stri_escape_unicode(.x)),
      floating = FALSE,include.rownames = include.rownames,add.to.row = addtorow,hline.after =
        c(-1)
    )
    cat("\\addtocounter{table}{-1}\n")
    cat(sprintf("\\label{%s}",label),'\n')
    if (landscape) {
      cat("\\end{center}\n \\end{landscape}")
    }else{
      cat("\\end{center}\n ")
    }
  }
##==================== Layouts ====================##
alMetCarF<-paste0('|l',#'|p{0.01\\textwidth}',#ID
           '|p{0.15\\textwidth}',#MetID
           '|p{0.35\\textwidth}',#Name
           '|r',                #Nflux5
           '|r',                #Nflux5
           '|r',                #Nflux5
           '|r',                #Nflux5
#           '|r',                #Nflux5
#           '|r',                #Nflux5
           '|r|'                 #,N
)

alEdgesOne<- paste0('|p{0.01\\textwidth}',#ID '|l',#
           '|p{0.15\\textwidth}',                #varname 
           '|p{0.15\\textwidth}',                #MetID.wt 
           '|r',                #original.wt 
           '|r',                #pval.wt 
           '|r',                #card.wt 
           '|r',                #pAct16 
           '|r',                #pAct32 
           '|r',                #pAct64 
           '|p{0.15\\textwidth}|'                #Name 
)

excelFileName<-"FBA-PRCC-EcoliK12.xlsx"
```
\pagebreak


# FBA Sensitivity Analysis

The main idea of FBA-PRCC analysis was introduced in our recent paper [@Sorokin:2023FBAPRCC]. 

```{r check.values, echo=FALSE,eval=TRUE}
#if(!exists('mdir')){mdir<-defDir}
fnums<-c(7,8,15,16,31,32,63,64)
mvar<-c("MetID","varname","Name","ExEquation","HMDF","ChEBI","InChI","SMILES","KEGG","SEED","BiGG","LipidMaps","MetaNetX","Reactome","PubChem")
```

## Escherichia coli str. K12 substr. MG1655

### Read data
We have calculated PRCC coefficients with different number of Sobol points 
for all exchange reactions for the metabolic model of 
*Escherichia coli str. K12 substr. MG1655* taken from the
AGORA2 collection [@Heinken:2023Genomescale]. The following code loads that data:

```{r read.rds.wt,eval=TRUE,cache=TRUE}
d<-"Escherichia_coli_str_K_12_substr_MG1655"
mdf<-fread(file=file.path('data',paste0(d,'.csv')))
fr64<-fread(file.path('data',d,'flux_range_64.csv'))[,-1]
lfdr<-list()
for(fnum in fnums){
    frdf<-fread(file = file.path('data',d,paste0('prcc_',fnum,'.csv')))
    lfdr[[sprintf('fnm%02d',fnum)]]<-frdf
}
df<-as.data.table(lfdr$fnm64)
```

```{r read.active.df16.wt,echo=FALSE,cache=FALSE,eval=TRUE}
dt<-fread(file=file.path('data',d,'fba_sobol_isactive.csv'))
idx<-match(names(dt)[-c(1,ncol(dt))],df$varname)
dtWT64<-dt
names(dtWT64)<-sub('^EX_','actEX_',names(dtWT64))
```

```{r calc.active.df16.wt,echo=FALSE,cache=TRUE,eval=TRUE}
df$PercActive16[idx]<-apply(as.data.frame(dt)[1:131072,-c(1,ncol(dt))],2,mean)
df$NActive16[idx]<-apply(as.data.frame(dt)[1:131072,-c(1,ncol(dt))],2,sum)
```
```{r calc.active.df32.wt,echo=FALSE,cache=TRUE,eval=TRUE}
df$PercActive32[idx]<-apply(as.data.frame(dt)[1:262144,-c(1,ncol(dt))],2,mean)
df$NActive32[idx]<-apply(as.data.frame(dt)[1:262144,-c(1,ncol(dt))],2,sum)
```

```{r calc.active.df64.wt,echo=FALSE,cache=TRUE,eval=TRUE}
df$PercActive64[idx]<-apply(as.data.frame(dt)[,-c(1,ncol(dt))],2,mean)
df$NActive64[idx]<-apply(as.data.frame(dt)[,-c(1,ncol(dt))],2,sum)
```

```{r filter.df.wt}
dfAll<-df
tmp<-as.data.frame(df)
dfw<-as.data.table(tmp)
```

### Estimate number of the Sobol point for PRCC value calculations
As was mentioned in [@Marino:2008methodology] we first need to check what number
of Sobol points is sufficient for reliable estimation of the PRCC values in the
space of as much as `r nrow(df)` dimensions. To do this we will calculate
Top-Down Concordance Coefficients (TDCC) between PRCC values calculated with different 
number of Sobol points and check how that value changes with that number. To make
graphs comparable we always keep the same difference in number of Sobol point used 
for TDCC calculations. 
Supplementary Figures \ref{fig:tdcc_8}, \ref{fig:tdcc_16}, \ref{fig:tdcc_32}, 
and \ref{fig:tdcc_64} shows that with increase in the number of Sobol points 
difference between PRCC coefficient decreases.

```{r prepare.tdcc.calculations}
sigUpset01<-list()
sigUpset05<-list()
sigUpset10<-list()
tdccVal<-list()
pvalDif<-list()
```

```{r tdcc.8.wt, fig.width=8,fig.height=8,fig.cap="\\label{fig:tdcc_8} Comparison of PRCC ranking calculated on 57344 (A) and 65536 (B) Sobol points."}
tdccVal[['fnm8']]<-tdcc(rbind(rank(lfdr$fnm07$original[!(is.na(lfdr$fnm07$T) & 
                                                             is.na(lfdr$fnm08$T))],
                                   ties.method = "random"),
                              rank(lfdr$fnm08$original[!(is.na(lfdr$fnm07$T) & 
                                                             is.na(lfdr$fnm08$T))],
                                   ties.method = "random")),plot=TRUE,pearson = TRUE)
n1<-7
n2<-8
prcc1<-lfdr[[sprintf('fnm%02d',n1)]]$original
prcc2<-lfdr[[sprintf('fnm%02d',n2)]]$original
p1<-length(prcc1)-1
p2<-length(prcc2)-1
N1<-8192 * as.numeric(n1)
N2<-8192 * as.numeric(n2)
z<-(normPRCC(prcc1)-normPRCC(prcc2))/sqrt(1/(N1-3-p1) + 1/(N2-3-p2))
pvalDif[[sprintf('fnm%02d',n2)]]<-data.frame(z=z,pval=pnorm(z),grp=sprintf('%06d-%06d',N1,N2),str='wt')
sigUpset01[[sprintf('fnm%06d',N2)]]<-
    lfdr[[sprintf('fnm%02d',n2)]]$varname[lfdr[[sprintf('fnm%02d',n2)]]$pval<=0.01]
sigUpset05[[sprintf('fnm%06d',N2)]]<-
    lfdr[[sprintf('fnm%02d',n2)]]$varname[lfdr[[sprintf('fnm%02d',n2)]]$pval<=0.05]
sigUpset10[[sprintf('fnm%06d',N2)]]<-
    lfdr[[sprintf('fnm%02d',n2)]]$varname[lfdr[[sprintf('fnm%02d',n2)]]$pval<=0.1]
```

```{r tdcc.16.wt, fig.width=8,fig.height=8,fig.cap="\\label{fig:tdcc_16} Comparison of PRCC ranking calculated on 122880 (A) and 131072 (B) Sobol points."}
tdccVal[['fnm16']]<-tdcc(rbind(
    rank(lfdr$fnm15$original[!(is.na(lfdr$fnm15$T) & is.na(lfdr$fnm16$T))],ties.method = "random"),
    rank(lfdr$fnm16$original[!(is.na(lfdr$fnm15$T) & is.na(lfdr$fnm16$T))],ties.method = "random")),
    plot=TRUE,pearson = TRUE)
n1<-15
n2<-16
prcc1<-lfdr[[sprintf('fnm%02d',n1)]]$original
prcc2<-lfdr[[sprintf('fnm%02d',n2)]]$original
p1<-length(prcc1)-1
p2<-length(prcc2)-1
N1<-8192 * n1
N2<-8192 * n2
z<-(normPRCC(prcc1)-normPRCC(prcc2))/sqrt(1/(N1-3-p1) + 1/(N2-3-p2))
pvalDif[[sprintf('fnm%02d',n2)]]<-
    data.frame(z=z,pval=pnorm(z),grp=sprintf('%06d-%06d',N1,N2),str='wt')
sigUpset01[[sprintf('fnm%06d',N2)]]<-
    lfdr[[sprintf('fnm%02d',n2)]]$varname[lfdr[[sprintf('fnm%02d',n2)]]$pval<=0.01]
sigUpset05[[sprintf('fnm%06d',N2)]]<-
    lfdr[[sprintf('fnm%02d',n2)]]$varname[lfdr[[sprintf('fnm%02d',n2)]]$pval<=0.05]
sigUpset10[[sprintf('fnm%06d',N2)]]<-
    lfdr[[sprintf('fnm%02d',n2)]]$varname[lfdr[[sprintf('fnm%02d',n2)]]$pval<=0.1]
```

```{r tdcc.32.wt, fig.width=8,fig.height=8,fig.cap="\\label{fig:tdcc_32} Comparison of PRCC ranking calculated on 253952 (A) and 262144 (B) Sobol points."}
tdccVal[['fnm32']]<-tdcc(rbind(
    rank(lfdr$fnm31$original[!(is.na(lfdr$fnm31$T) & is.na(lfdr$fnm32$T))],ties.method = "random"),
    rank(lfdr$fnm32$original[!(is.na(lfdr$fnm31$T) & is.na(lfdr$fnm32$T))],ties.method = "random")),
    plot=TRUE,pearson = TRUE)
n1<-31
n2<-32
prcc1<-lfdr[[sprintf('fnm%02d',n1)]]$original
prcc2<-lfdr[[sprintf('fnm%02d',n2)]]$original
p1<-length(prcc1)-1
p2<-length(prcc2)-1
N1<-8192 * n1
N2<-8192 * n2
z<-(normPRCC(prcc1)-normPRCC(prcc2))/sqrt(1/(N1-3-p1) + 1/(N2-3-p2))
pvalDif[[sprintf('fnm%02d',n2)]]<-data.frame(z=z,pval=pnorm(z),grp=sprintf('%06d-%06d',N1,N2),str='wt')
sigUpset01[[sprintf('fnm%06d',N2)]]<-
    lfdr[[sprintf('fnm%02d',n2)]]$varname[lfdr[[sprintf('fnm%02d',n2)]]$pval<=0.01]
sigUpset05[[sprintf('fnm%06d',N2)]]<-
    lfdr[[sprintf('fnm%02d',n2)]]$varname[lfdr[[sprintf('fnm%02d',n2)]]$pval<=0.05]
sigUpset10[[sprintf('fnm%06d',N2)]]<-
    lfdr[[sprintf('fnm%02d',n2)]]$varname[lfdr[[sprintf('fnm%02d',n2)]]$pval<=0.1]
```


```{r tdcc.64.wt, fig.width=8,fig.height=8,fig.cap="\\label{fig:tdcc_64} Comparison of PRCC ranking calculated on 516096 (A) and 524288 (B) Sobol points."}
tdccVal[['fnm64']]<-tdcc(rbind(
    rank(lfdr$fnm63$original[!(is.na(lfdr$fnm63$T) & is.na(lfdr$fnm64$T))],ties.method = "random"),
    rank(lfdr$fnm64$original[!(is.na(lfdr$fnm63$T) & is.na(lfdr$fnm64$T))],ties.method = "random")),
    plot=TRUE,pearson = TRUE)
n1<-63
n2<-64
prcc1<-lfdr[[sprintf('fnm%02d',n1)]]$original
prcc2<-lfdr[[sprintf('fnm%02d',n2)]]$original
p1<-length(prcc1)-1
p2<-length(prcc2)-1
N1<-8192 * n1
N2<-8192 * n2
z<-(normPRCC(prcc1)-normPRCC(prcc2))/sqrt(1/(N1-3-p1) + 1/(N2-3-p2))
pvalDif[[sprintf('fnm%02d',n2)]]<-data.frame(z=z,pval=pnorm(z),grp=sprintf('%06d-%06d',N1,N2),str='wt')
sigUpset01[[sprintf('fnm%06d',N2)]]<-lfdr[[sprintf('fnm%02d',n2)]]$varname[lfdr[[sprintf('fnm%02d',n2)]]$pval<=0.01]
sigUpset05[[sprintf('fnm%06d',N2)]]<-lfdr[[sprintf('fnm%02d',n2)]]$varname[lfdr[[sprintf('fnm%02d',n2)]]$pval<=0.05]
sigUpset10[[sprintf('fnm%06d',N2)]]<-lfdr[[sprintf('fnm%02d',n2)]]$varname[lfdr[[sprintf('fnm%02d',n2)]]$pval<=0.1]
```

The Supplementary Figure \ref{fig:tdcc_trend} and Supplementary Table \ref{tab:TDCC}
summarises data from Supplementary Figures \ref{fig:tdcc_8}, \ref{fig:tdcc_16}, 
\ref{fig:tdcc_32}, and \ref{fig:tdcc_64}. According to this results we have decided
that `r 64*8192` Sobol points is enough for reliable estimation of the PRCC values.

```{r tdcc.plot, fig.width=8,fig.height=8,fig.cap="\\label{fig:tdcc_trend} Top-Down Correlation Coefficient (TDCC) vs. number of Sobol points."}
tdccdf<-as.data.frame(do.call(rbind,tdccVal))
tdccdf$N<-c(8,16,32,64)*8192
qplot(N,pearson,data=tdccdf)+geom_line()+
    xlab('number of Sobol points')+
    theme_bw()
```

```{r tdcc.table,results='asis'}
printTable(tdccdf[,c('N','kendall','pearson')],
           'TDCC values at different number of Sobol points',digits = 3,
           landscape = FALSE,align = 'lrrr',include.rownames = TRUE,label = 'tab:TDCC')
```

Another method to estimate the number of Sobol points sufficient for the reliable PRCC value estimation, proposed in [@Marino:2008methodology] is to calculate the statistical significance 
of PRCC values calculated with different numbers of Sobol points. Supplementary 
Figures \ref{fig:pval_trend} and \ref{fig:lpval_trend} show how significance 
values change with the number of Sobol points. That data supports our choice of the
Sobol points number.

```{r pval.diff.plot, fig.width=8,fig.height=8,fig.cap="\\label{fig:pval_trend} Probability that difference of PRCC coefficients is random vs. number of Sobol points."}
pvalDF<-do.call(rbind,pvalDif)
p<-ggplot(pvalDF,aes(x=grp,y=pval))+geom_boxplot()+theme_bw()
print(p)
```

```{r pval.diff.plot.logy, fig.width=8,fig.height=8,fig.cap="\\label{fig:lpval_trend} Log-probability that difference of PRCC coefficients is random vs. number of Sobol points."}
p+scale_y_log10()+theme_bw()
```

Another parameter required for the following analysis is the p-value, which separates
significant PRCC values from PRCC values indistinguishable from zero. Supplementary
Figure \ref{fig:wt_prcc_pval} shows how PRCC significance is distributed. It could be
seen that distribution is bimodal with values grouped around zero and above 0.5.
According to this plot, the majority of PRCC values are not distinguishable from zero.

```{r hist.plot.wt,fig.width=8,fig.height=8,fig.cap="\\label{fig:wt_prcc_pval} Distribution of the significance of the exchange reaction boundaries PRCC values for the wild-type model."}
qplot(df$pval,bins = 50)+
    xlab('p-value')+
    ggtitle('Full dataset PRCC significance distribution for wild type')+
    theme_bw()
```

To choose the p-value threshold we compare sets of significant exchange reactions 
calculated with different numbers of Sobol points. Supplementary Figures 
\ref{fig:upset01}, \ref{fig:upset05}, and \ref{fig:upset10} show UpSet plots 
[@Lex:2014UpSet; @Conway:2017UpSetR] of 
intersection of exchange reaction sets significant at the p-values 1%, 5%, and 
10% correspondingly. It could be seen that the majority of sensitive reaction bounds 
are identified in all sizes of the data, and because large numbers of the bounds 
were found at p-value of 10% we will use this value in the following analysis.

```{r sig01.upset.plot, fig.width=8,fig.height=8,fig.cap="\\label{fig:upset01} Intersection of sets of fluxes, for which PRCC different from zero at the significance level of 1%, obtained with different number of Sobol points."}
upset(fromList(sigUpset01),nsets = length(sigUpset01))
```
```{r sig05.upset.plot, fig.width=8,fig.height=8,fig.cap="\\label{fig:upset05} Intersection of sets of fluxes, for which PRCC different from zero at the significance level of 5%, obtained with different number of Sobol points."}
upset(fromList(sigUpset05),nsets = length(sigUpset05))
```
```{r sig10.upset.plot, fig.width=8,fig.height=8,fig.cap="\\label{fig:upset10} Intersection of sets of fluxes, for which PRCC different from zero at the significance level of 10%, obtained with different number of Sobol points."}
upset(fromList(sigUpset10),nsets = length(sigUpset10))
```

### Histidine and tryptophan flux plots

We are going to compare the sensitivity profile of the wild-type model with two auxotrophic
mutants $\Delta hisD$ and $\Delta trpB$, which are unable to produce internally
histidine and tryptophan correspondently. 

Supplementary Figures \ref{fig:wt_his_bs}, \ref{fig:wt_his_bs_f}, \ref{fig:wt_his_bf}, and \ref{fig:wt_his_bf_f} demonstrate how histidine influx bound value influences
histidine influx itself and biomass production.

```{r read.wt.flux.plot.dataset}
fluxes_WT<-fread(file=file.path('data',d,'mutant_related_fluxes.csv'))
```
```{r prepare.wt.his.plot.dataset}
re_names<-c("ALAHIS1c", "EX_his_L(e)", "HISCAT1pp", "HIS_Ltex", "HISabcpp", "HISt2pp","EX_alahis(e)")
hisL<-fluxes_WT[,match(paste0(re_names,'_fl'),names(fluxes_WT)),with=FALSE]
pltDFwtHis<-data.frame(Solution=fluxes_WT$Solution,
                  influx_HisL=-1*(hisL$`EX_his_L(e)_fl`),
                  inBound_HisL=1000*fluxes_WT$`EX_his_L(e)_l`,
                  outBound_HisL=1000*fluxes_WT$`EX_his_L(e)_u`,
                  activityInflux_HisL=factor(dtWT64$`actEX_his_L(e)_l`,
                                        levels = c(1,0),labels = c('active','inactive')),
                  activityOutflux_HisL=factor(dtWT64$`actEX_his_L(e)_u`,
                                         levels = c(1,0),labels = c('active','inactive')),
                  influx_AlaHis=-1*(hisL$`EX_alahis(e)_fl`),
                  inBound_AlaHis=1000*fluxes_WT$`EX_alahis(e)_l`,
                  outBound_AlaHis=1000*fluxes_WT$`EX_alahis(e)_u`,
                  activityInflux_AlaHis=factor(dtWT64$`actEX_alahis(e)_l`,
                                        levels = c(1,0),labels = c('active','inactive')),
                  activityOutflux_AlaHis=factor(dtWT64$`actEX_alahis(e)_u`,
                                         levels = c(1,0),labels = c('active','inactive')))
```

```{r plot.wt.his.bndvssol,fig.format='png',fig.dpi=600,fig.cap="\\label{fig:wt_his_bs} Influence of the histidine influx boundary on the biomass production in the wild-type. Sobol points where the boundary was active shown in red."}
qplot(inBound_HisL,Solution,data = pltDFwtHis,color=activityInflux_HisL,log='xy')+
    #ggtitle(main =bquote("Histidine influx boundary in wildtype"))+
    theme_bw()
```

```{r plot.wt.his.bndvssol.facet,fig.format='png',fig.dpi=600,fig.cap="\\label{fig:wt_his_bs_f} Influence of the histidine influx boundary on the biomass production in the wild-type. Sobol points where the boundary was active shown in red and placed on different panels."}
qplot(inBound_HisL,Solution,data = pltDFwtHis,color=activityInflux_HisL,log='xy',
      facets = activityInflux_HisL ~ .)+
    #ggtitle(bquote("Histidine influx boundary in wildtype"))+
    theme_bw()
```

```{r plot.wt.his.bndvsinflux,fig.format='png',fig.dpi=600,fig.cap="\\label{fig:wt_his_bf} Influence of the histidine influx boundary on the actual histidine influx value in the wild-type. Sobol points where the boundary was active shown in red."}
qplot(inBound_HisL,influx_HisL,data = pltDFwtHis,color = activityInflux_HisL,log='xy')+
    #ggtitle(bquote("Histidine influx in wildtype"))+
    theme_bw()
```

```{r plot.wt.his.bndvsinflux_facet,fig.format='png',fig.dpi=600,fig.cap="\\label{fig:wt_his_bf_f} Influence of the histidine influx boundary on the actual histidine influx value in the wild-type. Sobol points where the boundary was active shown in red and placed on different panels."}
qplot(inBound_HisL,influx_HisL,data = pltDFwtHis,color = activityInflux_HisL,log='xy',
      facets = activityInflux_HisL ~ .)+
    #ggtitle(bquote("Histidine influx in wildtype"))+
    theme_bw()
```



```{r plot.wt.alahis.bndvssol,fig.format='png',fig.dpi=600,fig.cap="\\label{fig:wt_alahis_bs} Influence of the AlaHis influx boundary on the biomass production in the wild-type. Sobol points where the boundary was active shown in red."}
qplot(inBound_AlaHis,Solution,data = pltDFwtHis,color=activityInflux_AlaHis,log='xy')+
    #ggtitle(main =bquote("Histidine influx boundary in wildtype"))+
    theme_bw()
```

```{r plot.wt.alahis.bndvssol.facet,fig.format='png',fig.dpi=600,fig.cap="\\label{fig:wt_alahis_bs_f} Influence of the AlaHis influx boundary on the biomass production in the wild-type. Sobol points where the boundary was active shown in red and placed on different panels."}
qplot(inBound_AlaHis,Solution,data = pltDFwtHis,color=activityInflux_AlaHis,log='xy',
      facets = activityInflux_AlaHis ~ .)+
    #ggtitle(bquote("Histidine influx boundary in wildtype"))+
    theme_bw()
```

```{r plot.wt.alahis.bndvsinflux,fig.format='png',fig.dpi=600,fig.cap="\\label{fig:wt_alahis_bf} Influence of the AlaHis influx boundary on the actual AlaHis influx value in the wild-type. Sobol points where the boundary was active shown in red."}
qplot(inBound_AlaHis,influx_AlaHis,data = pltDFwtHis,color = activityInflux_AlaHis,log='xy')+
    #ggtitle(bquote("Histidine influx in wildtype"))+
    theme_bw()
```

```{r plot.wt.alahis.bndvsinflux_facet,fig.format='png',fig.dpi=600,fig.cap="\\label{fig:wt_alahis_bf_f} Influence of the AlaHis influx boundary on the actual AlaHis influx value in the wild-type. Sobol points where the boundary was active shown in red and placed on different panels."}
qplot(inBound_AlaHis,influx_AlaHis,data = pltDFwtHis,color = activityInflux_AlaHis,log='xy',
      facets = activityInflux_AlaHis ~ .)+
    #ggtitle(bquote("Histidine influx in wildtype"))+
    theme_bw()
```

Supplementary Figures \ref{fig:wt_trp_bs}, \ref{fig:wt_trp_bs_f}, \ref{fig:wt_trp_bf}, and \ref{fig:wt_trp_bf_f} demonstrate how tryptophan influx bound value influences
tryptophan influx itself and biomass production.

```{r prepare.wt.trpB.plot.dataset}
re_names<-c("EX_trp_L(e)", "TRPAS2i", "TRPTA", "TRP_Ltex", "TRPt2rpp")
trpB<-fluxes_WT[,match(paste0(re_names,'_fl'),names(fluxes_WT)),with=FALSE]
pltDFwtTrp<-data.frame(Solution=fluxes_WT$Solution,
                  influx_TrpL=-1*(trpB$`EX_trp_L(e)_fl`),
                  inBound_TrpL=1000*fluxes_WT$`EX_trp_L(e)_l`,
                  outBound_TrpL=1000*fluxes_WT$`EX_trp_L(e)_u`,
                  activityInflux=factor(dtWT64$`actEX_trp_L(e)_l`,levels = c(1,0),labels = c('active','inactive')),
                  activityOutflux=factor(dtWT64$`actEX_trp_L(e)_u`,levels = c(1,0),labels = c('active','inactive')))
```
```{r plot.wt.trpB.bndvssol,fig.format='png',fig.dpi=600,fig.dpi=600,fig.cap="\\label{fig:wt_trp_bs_f} Influence of the tryptophan influx boundary on the biomass production in the wild-type. Sobol points where the boundary was active shown in red and placed on different panels."}
qplot(inBound_TrpL,Solution,data = pltDFwtTrp,color=activityInflux,log='xy')+
    #ggtitle(bquote("Tryptophan influx boundary in wildtype"))+
    theme_bw()
```
```{r plot.wt.trpB.bndvssol_facet,fig.format='png',fig.dpi=600,fig.cap="\\label{fig:wt_trp_bs} Influence of the tryptophan influx boundary on the biomass production in the wild-type. Sobol points where the boundary was active shown in red."}
qplot(inBound_TrpL,Solution,data = pltDFwtTrp,color=activityInflux,log='xy',
      facets = activityInflux ~ .)+
    #ggtitle(bquote("Tryptophan influx boundary in wildtype"))+
    theme_bw()
```

```{r plot.wt.trpB.bndvsinflu,fig.format='png',fig.dpi=600,fig.cap="\\label{fig:wt_trp_bf} Influence of the tryptophan influx boundary on the actual tryptophan influx value in the wild-type. Sobol points where the boundary was active shown in red."}
qplot(inBound_TrpL,influx_TrpL,data = pltDFwtTrp,color = activityInflux,log='x')+
    #ggtitle(bquote("Tryptophan influx in wildtype"))+
    theme_bw()
```

```{r plot.wt.trpB.bndvsinflu_facet,fig.format='png',fig.dpi=600,fig.cap="\\label{fig:wt_trp_bf_f} Influence of the tryptophan influx boundary on the actual tryptophan influx value in the wild-type. Sobol points where the boundary was active shown in red and placed on different panels."}
qplot(inBound_TrpL,influx_TrpL,data = pltDFwtTrp,color = activityInflux,log='x',
      facets = activityInflux ~ .)+
    #ggtitle(bquote("Tryptophan influx in wildtype"))+
    theme_bw()
```

```{r flux.sign.wt}
t<-as.data.frame(table(sign(pltDFwtHis$influx_AlaHis)))
names(t)<-c('Flux direction','Count')
t$Flux<-'AlaHis'
signDF<-t
t<-as.data.frame(table(sign(pltDFwtHis$influx_HisL)))
names(t)<-c('Flux direction','Count')
t$Flux<-'HisL'
signDF<-rbind(signDF,t)
t<-as.data.frame(table(sign(pltDFwtTrp$influx_TrpL)))
names(t)<-c('Flux direction','Count')
t$Flux<-'TrpL'
signDF<-rbind(signDF,t)
signDF$Percent<-signDF$Count*100/nrow(pltDFwtTrp)
levels(signDF$`Flux direction`)<-c('Out','Blocked','In')
signDF<-signDF[,c('Flux','Flux direction','Count','Percent')]
pander(signDF,
       caption=paste('\\label{tab:flx_sign_wt} Fractions of Sobol points groupped ',
                     'according to direction of exchange fluxes for AlaHis, HisL, ',
                     'and TrpL.'))
```


### Interaction type table

We group all significantly sensitive exchange reaction boundary values and 
corresponding metabolites into four groups according to the type of boundary 
(upper or lower) and sign of the significance coefficient (PRCC). Number of
reactions in each group is shown in the Supplementary Table \ref{tab:wt_int_fq}.
Some metabolites could have both their exchange reactions bound significantly,
such metabolites are listed in the Supplementary Table \ref{tab:bidir_met_p10}. 

Metabolites, which exchange reaction lower or upper boundary have a positive sensitivity coefficient, 
we will call substrates and products correspondingly. Substrates and products for the
wild-type model are shown in Supplementary Tables \ref{tab:SubsWT10} and \ref{tab:ProdsWT10}.

Metabolites, which exchange reaction lower or upper boundary have a negative sensitivity coefficient, 
we will call repellents and attractants correspondingly. Repellents and attractants for the
wild-type model are shown in Supplementary Tables \ref{tab:RepsWT10} and \ref{tab:AttrsWT10}.

```{r interaction.types.wt}
prnTbl<-df[(card>1),c(1:3,6,32,33,35,37,9)]
prnTbl10<-prnTbl[(pval<=0.1)]
pander(table(sign(prnTbl10$original),sub('.+_(.+)$','\\1',prnTbl10$varname)),
       caption = '\\label{tab:wt_int_fq} Sign/bound frequency table for wild-type model')
```

#### Substrate table
```{r prepare.substr.table.wt}
subTbl<-prnTbl[grepl('.+_l$',varname)&(original>0)]
subTbl10<-subTbl[(pval<=0.1)]
```

```{r substr10.met.car.table.wt,results='asis',cache=FALSE}
printTable(subTbl10[order(pval,-1*original)], digits = 5,
           main = 'Wild-type substrates',
           landscape = TRUE,align = alEdgesOne,
           label = 'tab:SubsWT10')
```

#### Product table
```{r prepare.prd.table.wt}
prdTbl<-prnTbl[grepl('.+_u$',varname)&(original>0)]
prdTbl10<-prdTbl[(pval<=0.1)]
```

```{r prd10.met.car.table.wt,results='asis',cache=FALSE}
printTable(prdTbl10[order(pval,-1*original)], digits = 5,
           main = 'Wild-type products',
           landscape = TRUE,align = alEdgesOne,
           label = 'tab:ProdsWT10')
```

#### Attranctant table
```{r prepare.attr.table.wt}
attrTbl<-prnTbl[grepl('.+_u$',varname)&(original<0)]
attrTbl10<-attrTbl[(pval<=0.1)]
```

```{r attr10.met.car.table.wt,results='asis',cache=FALSE}
printTable(attrTbl10[order(pval,original)], digits = 5,
           main = 'Wild-type attractants',
           landscape = TRUE,align = alEdgesOne,
           label = 'tab:AttrsWT10')
```


#### Repellent table
```{r prepare.rep.table.wt}
repTbl<-prnTbl[grepl('.+_l$',varname)&(original<0)]
repTbl10<-repTbl[(pval<=0.1)]
```

```{r rep10.met.car.table.wt,results='asis',cache=FALSE}
printTable(repTbl10[order(pval,original)], digits = 5,
           main = 'Wild-type repellents',
           landscape = TRUE,align = alEdgesOne,
           label = 'tab:RepsWT10')
```

## Escherichia coli str. K12 substr. MG1655 ko hisD

Auxotrophic mutant Escherichia coli str. K12 substr. MG1655 $\Delta hisD$ was 
created by knocking out hisD gene, which encodes **Histidinol dehydrogenase**. 
This enzyme controls the last step of the *de novo* histidine biosynthesis. 
The original AGORA2 model has a discrepancy in gene annotation of the *de novo* 
histidine biosynthesis. Reactions *HISTLOX* and *HISTDOX* are controlled by the gene `511145.12.peg.2097`, while reaction *HISTD* is controlled by gene 
named `511145.12.2097_15.peg`, which do not exist in the PATRIC database. Because
of this mutant model was created by removing reactions  *HISTLOX*, *HISTD*, and *HISTDOX*.

### Read data

As we already identify the desired number of Sobol point by analysis of the 
wild type model, we load only one set of data. The following code loads simulation data:
```{r read.rds.hisD,eval=TRUE,cache=TRUE}
d<-"Escherichia_coli_str_K_12_substr_MG1655_ko_hisD"
df<-fread(file = file.path('data',d,'prcc_64.csv'))
```

```{r read.active.df16.hisD,echo=FALSE,cache=FALSE,eval=TRUE}
dt<-fread(file=file.path('data',d,'fba_sobol_isactive.csv'))
dtHisD64<-dt
names(dtHisD64)<-sub('^EX_','actEX_',names(dtHisD64))
idx<-match(names(dt)[-c(1,ncol(dt))],df$varname)
```

```{r calc.active.df16.hisD,echo=FALSE,cache=TRUE,eval=TRUE}
df$PercActive16[idx]<-apply(as.data.frame(dt)[1:131072,-c(1,ncol(dt))],2,mean)
df$NActive16[idx]<-apply(as.data.frame(dt)[1:131072,-c(1,ncol(dt))],2,sum)
```
```{r calc.active.df32.hisD,echo=FALSE,cache=TRUE,eval=TRUE}
df$PercActive32[idx]<-apply(as.data.frame(dt)[1:262144,-c(1,ncol(dt))],2,mean)
df$NActive32[idx]<-apply(as.data.frame(dt)[1:262144,-c(1,ncol(dt))],2,sum)
```

```{r calc.active.df64.hisD,echo=FALSE,cache=TRUE,eval=TRUE}
df$PercActive64[idx]<-apply(as.data.frame(dt)[,-c(1,ncol(dt))],2,mean)
df$NActive64[idx]<-apply(as.data.frame(dt)[,-c(1,ncol(dt))],2,sum)
```

```{r filter.df.hisD}
dfAll<-rbind(dfAll,df,fill=TRUE)
dfw<-merge(dfw,df,by=mvar,all.x=TRUE,suffixes = c('.wt',''))
```
```{r hist.plot.hisD,fig.width=8,fig.height=8,fig.cap="\\label{fig:hisD_prcc_pval} Distribution of the significance of the exchange reaction boundaries PRCC values for the $\\Delta hisD$ model."}
qplot(df$pval,bins = 50)+
    xlab('p-value')+
    #ggtitle('Full dataset PRCC significance distribution for hisD')+
    theme_bw()
```

### Histidine flux plots

Similar to Supplementary Figures \ref{fig:wt_his_bs}, \ref{fig:wt_his_bs_f}, \ref{fig:wt_his_bf}, and \ref{fig:wt_his_bf_f}, which show influence of the histidine influx on the wild-type model, Supplementary Figures \ref{fig:hisD_his_bs}, \ref{fig:hisD_his_bs_f}, \ref{fig:hisD_his_bf}, and \ref{fig:hisD_his_bf_f} show influence of the histidine influx on the $\Delta hisD$ model.
```{r prepare.hisD.plot.dataset}
fluxes_hisD<-fread(file=file.path('data',d,'mutant_related_fluxes.csv'))
re_names<-c("ALAHIS1c", "EX_his_L(e)", "HISCAT1pp", "HIS_Ltex", "HISabcpp", "HISt2pp","EX_alahis(e)")
hisL<-fluxes_hisD[,match(paste0(re_names,'_fl'),names(fluxes_hisD)),with=FALSE]
pltDFhis<-data.frame(Solution=fluxes_hisD$Solution,
                  influx_HisL=-1*(hisL$`EX_his_L(e)_fl`),
                  inBound_HisL=1000*fluxes_hisD$`EX_his_L(e)_l`,
                  outBound_HisL=1000*fluxes_hisD$`EX_his_L(e)_u`,
                  activityInflux_HisL=factor(dtHisD64$`actEX_his_L(e)_l`,
                                        levels = c(1,0),labels = c('active','inactive')),
                  activityOutflux_HisL=factor(dtHisD64$`actEX_his_L(e)_u`,
                                         levels = c(1,0),labels = c('active','inactive')),
                  influx_AlaHis=-1*(hisL$`EX_alahis(e)_fl`),
                  inBound_AlaHis=1000*fluxes_hisD$`EX_alahis(e)_l`,
                  outBound_AlaHis=1000*fluxes_hisD$`EX_alahis(e)_u`,
                  activityInflux_AlaHis=factor(dtHisD64$`actEX_alahis(e)_l`,
                                        levels = c(1,0),labels = c('active','inactive')),
                  activityOutflux_AlaHis=factor(dtHisD64$`actEX_alahis(e)_u`,
                                         levels = c(1,0),labels = c('active','inactive')))

```
```{r plot.hisD.bndvssol,fig.format='png',fig.dpi=600,fig.cap="\\label{fig:hisD_his_bs} Influence of the histidine influx boundary on the biomass production in the $\\Delta hisD$ model. Sobol points where the boundary was active shown in red."}
qplot(inBound_HisL,Solution,data = pltDFhis,color=activityInflux_HisL,log='xy')+
    #ggtitle(bquote("Histidine influx boundary in "~Delta~"hisD mutant"))+
    theme_bw()
```
```{r plot.hisD.bndvssol_facet,fig.format='png',fig.dpi=600,fig.cap="\\label{fig:hisD_his_bs_f} Influence of the histidine influx boundary on the biomass production in the $\\Delta hisD$ model. Sobol points where the boundary was active shown in red and placed on different panels."}
qplot(inBound_HisL,Solution,data = pltDFhis,color=activityInflux_HisL,log='xy',
      facets = activityInflux_HisL ~ .)+
    #ggtitle(bquote("Histidine influx boundary in "~Delta~"hisD mutant"))+
    theme_bw()
```

```{r plot.hisD.bndvsinflux,fig.format='png',fig.dpi=600,fig.cap="\\label{fig:hisD_his_bf} Influence of the histidine influx boundary on the actual histidine influx value in the $\\Delta hisD$ model. Sobol points where the boundary was active shown in red."}
qplot(inBound_HisL,influx_HisL,data = pltDFhis,color = activityInflux_HisL,
      log='xy')+
    #ggtitle(bquote("Histidine influx in "~Delta~"hisD mutant"))+
    theme_bw()
```

```{r plot.hisD.bndvsinflux_facet,fig.format='png',fig.dpi=600,fig.cap="\\label{fig:hisD_his_bf_f} Influence of the histidine influx boundary on the actual histidine influx value in the $\\Delta hisD$ model. Sobol points where the boundary was active shown in red and placed on different panels."}
qplot(inBound_HisL,influx_HisL,data = pltDFhis,color = activityInflux_HisL,log='xy',
      facets = activityInflux_HisL ~ .)+
    #ggtitle(bquote("Histidine influx in "~Delta~"hisD mutant"))+
    theme_bw()
```

```{r plot.hisD.bndvssol_AlaHis,fig.format='png',fig.dpi=600,fig.cap="\\label{fig:hisD_alahis_bs} Influence of the AlaHis influx boundary on the biomass production in the $\\Delta hisD$ model. Sobol points where the boundary was active shown in red."}
qplot(inBound_AlaHis,Solution,data = pltDFhis,color=activityInflux_AlaHis,log='xy')+
    #ggtitle(bquote("AlaHis influx boundary in "~Delta~"hisD mutant"))+
    theme_bw()
```
```{r plot.hisD.bndvssol_AlaHis_facet,fig.format='png',fig.dpi=600,fig.cap="\\label{fig:hisD_alahis_bs_f} Influence of the AlaHis influx boundary on the biomass production in the $\\Delta hisD$ model. Sobol points where the boundary was active shown in red and placed on different panels."}
qplot(inBound_AlaHis,Solution,data = pltDFhis,color=activityInflux_AlaHis,log='xy',
      facets = activityInflux_AlaHis ~ .)+
    #ggtitle(bquote("Histidine influx boundary in "~Delta~"hisD mutant"))+
    theme_bw()
```

```{r plot.hisD.bndvsinflux_AlaHis,fig.format='png',fig.dpi=600,fig.cap="\\label{fig:hisD_alahis_bf} Influence of the AlaHis influx boundary on the actual AlaHis influx value in the $\\Delta hisD$ model. Sobol points where the boundary was active shown in red."}
qplot(inBound_AlaHis,influx_AlaHis,data = pltDFhis,color = activityInflux_AlaHis,
      log='xy')+
    #ggtitle(bquote("Histidine influx in "~Delta~"hisD mutant"))+
    theme_bw()
```

```{r plot.hisD.bndvsinflux_AlaHis_facet,fig.format='png',fig.dpi=600,fig.cap="\\label{fig:hisD_alahis_bf_f} Influence of the AlaHis influx boundary on the actual AlaHis influx value in the $\\Delta hisD$ model. Sobol points where the boundary was active shown in red and placed on different panels."}
qplot(inBound_AlaHis,influx_AlaHis,data = pltDFhis,color = activityInflux_AlaHis,log='xy',
      facets = activityInflux_AlaHis ~ .)+
    #ggtitle(bquote("Histidine influx in "~Delta~"hisD mutant"))+
    theme_bw()
```



### Interaction type table

We group all significantly sensitive exchange reaction boundary values and 
corresponding metabolites into four groups according to the type of boundary 
(upper or lower) and sign of the significance coefficient (PRCC). The number of
reactions in each group for $\Delta hisD$ is shown in the Supplementary Table \ref{tab:hisD_int_fq}.

Metabolites, which exchange reaction lower or upper boundary have a positive sensitivity coefficient, 
we will call substrates and products correspondingly. Substrates and products for the
$\Delta hisD$ model shown in Supplementary Tables \ref{tab:SubsHisD10} and \ref{tab:ProdsHisD10}.

Metabolites, which exchange reaction lower or upper boundary have a negative sensitivity coefficient, 
we will call repellents and attractants correspondingly. Repellents and attractants for the
$\Delta hisD$ model shown in Supplementary Tables \ref{tab:RepsHisD10} and \ref{tab:AttrsHisD10}.

```{r interaction.types.hisD}
prnTbl<-df[(card>1),c(1:3,6,31,32,34,36,8)]
prnTbl10<-prnTbl[(pval<=0.1)]
pander(table(sign(prnTbl10$original),sub('.+_(.+)$','\\1',prnTbl10$varname)),
       caption = '\\label{tab:hisD_int_fq} Sign/bound frequency table for $\\Delta hisD$ model')
```

#### Substrate table
```{r prepare.substr.table.hisD}
subTbl<-prnTbl[grepl('.+_l$',varname)&(original>0)]
subTbl10<-subTbl[(pval<=0.1)]
```

```{r substr10.met.car.table.hisD,results='asis',cache=FALSE}
printTable(subTbl10[order(pval,-1*original)], digits = 5,
           main = '$\\Delta hisD$ substrates',
           landscape = TRUE,align = alEdgesOne,
           label = 'tab:SubsHisD10')
```

#### Product table
```{r prepare.prd.table.hisD}
prdTbl<-prnTbl[grepl('.+_u$',varname)&(original>0)]
prdTbl10<-prdTbl[(pval<=0.1)]
```

```{r prd10.met.car.table.hisD,results='asis',cache=FALSE}
printTable(prdTbl10[order(pval,-1*original)], digits = 5,
           main = '$\\Delta hisD$ products',
           landscape = TRUE,align = alEdgesOne,
           label = 'tab:ProdsHisD10')
```

#### Attranctant table
```{r prepare.attr.table.hisD}
attrTbl<-prnTbl[grepl('.+_u$',varname)&(original<0)]
attrTbl10<-attrTbl[(pval<=0.1)]
```

```{r attr10.met.car.table.hisD,results='asis',cache=FALSE}
printTable(attrTbl10[order(pval,original)], digits = 5,
           main = '$\\Delta hisD$ attractants',
           landscape = TRUE,align = alEdgesOne,
           label = 'tab:AttrsHisD10')
```

#### Repellent table
```{r prepare.rep.table.hisD}
repTbl<-prnTbl[grepl('.+_l$',varname)&(original<0)]
repTbl10<-repTbl[(pval<=0.1)]
```

```{r rep10.met.car.table.hisD,results='asis',cache=FALSE}
printTable(repTbl10[order(pval,original)], digits = 5,
           main = '$\\Delta hisD$ repellents',
           landscape = TRUE,align = alEdgesOne,
           label = 'tab:RepsHisD10')
```

## Escherichia coli str. K12 substr. MG1655 ko hisD no AlaHis transport

Analysis of the simulation results of $\Delta hisD$ model demonstrates unexpected
results: the model appears to be insensitive to the histidine influx. Detailed analysis
of the model structure reveals that histidine could be provided not only as pure
amino acid but also as dipeptide AlaHis. To exclude the influence of the dipeptide from
analysis new model was created by removing the reaction *ALAHIStex* to prevent
AlaHis assimilation. The same result could be obtained by removing the exchange reaction
*EX_alahis(e)*, but that would change the dimensionality of the parameter space 
sampled by the Sobol points. This change will make it impossible to compare
flux distributions between this model and other variants. 

#### Read data

The following code loads simulation data:
```{r read.rds.hisD.noalahis,eval=TRUE,cache=TRUE}
d<-"Escherichia_coli_str_K_12_substr_MG1655_ko_hisD_block_alahis"
df<-fread(file = file.path('data',d,'prcc_64.csv'))
```

```{r read.active.df16.hisD.noalahis,echo=FALSE,cache=FALSE,eval=TRUE}
dt<-fread(file=file.path('data',d,'fba_sobol_isactive.csv'))
dtHisD64_alahis<-dt
names(dtHisD64_alahis)<-sub('^EX_','actEX_',names(dtHisD64_alahis))
idx<-match(names(dt)[-c(1,ncol(dt))],df$varname)
```


```{r calc.active.df16.hisD.noalahis,echo=FALSE,cache=TRUE,eval=TRUE}
df$PercActive16[idx]<-apply(as.data.frame(dt)[1:131072,-c(1,ncol(dt))],2,mean)
df$NActive16[idx]<-apply(as.data.frame(dt)[1:131072,-c(1,ncol(dt))],2,sum)
```
```{r calc.active.df32.hisD.noalahis,echo=FALSE,cache=TRUE,eval=TRUE}
df$PercActive32[idx]<-apply(as.data.frame(dt)[1:262144,-c(1,ncol(dt))],2,mean)
df$NActive32[idx]<-apply(as.data.frame(dt)[1:262144,-c(1,ncol(dt))],2,sum)
```

```{r calc.active.df64.hisD.noalahis,echo=FALSE,cache=TRUE,eval=TRUE}
df$PercActive64[idx]<-apply(as.data.frame(dt)[,-c(1,ncol(dt))],2,mean)
df$NActive64[idx]<-apply(as.data.frame(dt)[,-c(1,ncol(dt))],2,sum)
```


```{r filter.df.hisD.noalahis}
dfAll<-rbind(dfAll,df,fill=TRUE )
dfw<-merge(dfw,df,by=mvar,all.x=TRUE,suffixes = c('.hisD',''))
```


```{r hist.plot.hisD.noalahis,fig.width=8,fig.height=8,fig.cap="\\label{fig:alahis_prcc_pval} Distribution of the significance of the exchange reaction boundaries PRCC values for the $\\Delta hisD$ model with blocked assimilation of the AlaHis dipeptide."}
qplot(df$pval)+
    xlab('p-value')+
    theme_bw()#+
    #ggtitle('Full dataset PRCC significance distribution for hisD with AlaHis blocked')
```

### Histidine flux plots

Similar to Supplementary Figures \ref{fig:wt_his_bs}, \ref{fig:wt_his_bs_f}, 
\ref{fig:wt_his_bf}, and \ref{fig:wt_his_bf_f}, which show influence of the 
histidine influx on the wild-type model, Supplementary Figures \ref{fig:hisD2_his_bs}, 
\ref{fig:hisD2_his_bs_f}, \ref{fig:hisD2_his_bf}, and \ref{fig:hisD2_his_bf_f} 
show the influence of the histidine influx on the $\Delta hisD$ model  with 
blocked AlaHis assimilation.

```{r prepare.hisD.noalahis.plot.dataset}
fluxes_hisD_alahis<-fread(file=file.path('data',d,'mutant_related_fluxes.csv'))
re_names<-c("ALAHIS1c", "EX_his_L(e)", "HISCAT1pp", "HIS_Ltex", "HISabcpp", "HISt2pp")
hisL<-fluxes_hisD_alahis[,match(paste0(re_names,'_fl'),names(fluxes_hisD)),with=FALSE]
pltDFhis<-data.frame(Solution=fluxes_hisD_alahis$Solution,
                  influx_HisL=-1*(hisL$`EX_his_L(e)_fl`),
                  inBound_HisL=1000*fluxes_hisD$`EX_his_L(e)_l`,
                  outBound_HisL=1000*fluxes_hisD$`EX_his_L(e)_u`,
                  activityInflux=factor(dtHisD64_alahis$`actEX_his_L(e)_l`,levels = c(1,0),labels = c('active','inactive')),
                  activityOutflux=factor(dtHisD64_alahis$`actEX_his_L(e)_u`,levels = c(1,0),labels = c('active','inactive')))

```
```{r plot.hisD.noalahis.bndvssol,fig.format='png',fig.dpi=600,fig.cap="\\label{fig:hisD2_his_bs} Influence of the histidine influx boundary on the biomass production in the $\\Delta hisD$ model with blocked AlaHis assimilation. Sobol points where the boundary was active shown in red."}
qplot(inBound_HisL,Solution,data = pltDFhis,color=activityInflux,log='xy')+
    #ggtitle(bquote("Histidine influx boundary in "~Delta~"hisD mutant"))+
    theme_bw()
```
```{r plot.hisD.noalahis.bndvssol.facet,fig.format='png',fig.dpi=600,fig.cap="\\label{fig:hisD2_his_bs_f} Influence of the histidine influx boundary on the biomass production in the $\\Delta hisD$ model with blocked AlaHis assimilation. Sobol points where the boundary was active shown in red and placed on different panels."}
qplot(inBound_HisL,Solution,data = pltDFhis,color=activityInflux,log='xy',
      facets = activityInflux ~ .)+
    #ggtitle(bquote("Histidine influx boundary in "~Delta~"hisD mutant"))+
    theme_bw()
```

```{r plot.hisD.noalahis.bndvsinflux,fig.format='png',fig.dpi=600,fig.cap="\\label{fig:hisD2_his_bf} Influence of the histidine influx boundary on the actual histidine influx value in the $\\Delta hisD$ model with blocked AlaHis assimilation. Sobol points where the boundary was active shown in red."}
qplot(inBound_HisL,influx_HisL,data = pltDFhis,color = activityInflux,log='xy')+
    #ggtitle(bquote("Histidine influx in "~Delta~"hisD mutant"))+
    theme_bw()
```

```{r plot.hisD.noalahis.bndvsinflux_facet,fig.format='png',fig.dpi=600,fig.cap="\\label{fig:hisD2_his_bf_f} Influence of the histidine influx boundary on the actual histidine influx value in the $\\Delta hisD$ model with blocked AlaHis assimilation. Sobol points where the boundary was active shown in red and placed on different panels."}
qplot(inBound_HisL,influx_HisL,data = pltDFhis,color = activityInflux,log='xy',
      facets = activityInflux ~ .)+
    #ggtitle(bquote("Histidine influx in "~Delta~"hisD mutant"))+
    theme_bw()
```



### Interaction type table

We group all significantly sensitive exchange reaction boundary values and 
corresponding metabolites into four groups according to the type of boundary 
(upper or lower) and sign of the significance coefficient (PRCC). The number of
reactions in each group for $\Delta hisD$ is shown in the Supplementary Table \ref{tab:hisD2_int_fq}.

Metabolites, which exchange reaction lower or upper boundary have a positive sensitivity coefficient, 
we will call substrates and products correspondingly. Substrates and products for the
$\Delta hisD$ model shown in Supplementary Tables \ref{tab:SubsHisDnoalahis10} 
and \ref{tab:ProdsHisDnoalahis10}.

Metabolites, which exchange reaction lower or upper boundary have a negative sensitivity coefficient, 
we will call repellents and attractants correspondingly. Repellents and attractants for the
$\Delta hisD$ model shown in Supplementary Tables \ref{tab:RepsHisDnoalahis10} 
and \ref{tab:AttrsHisDnoalahis10}.

```{r interaction.types.hisD.noalahis}
prnTbl<-df[(card>1),c(1:3,6,31,32,34,36,8)]
#prnTbl<-df[(card>1),c(1:3,6,8)]
prnTbl10<-prnTbl[(pval<=0.1)]
pander(table(sign(prnTbl10$original),sub('.+_(.+)$','\\1',prnTbl10$varname)),
       caption = '\\label{tab:hisD2_int_fq} Sign/bound frequency table for $\\Delta hisD$ model with blocked AlaHis assimilation.')
```

#### Substrate table
```{r prepare.substr.table.hisD.noalahis}
subTbl<-prnTbl[grepl('.+_l$',varname)&(original>0)]
subTbl10<-subTbl[(pval<=0.1)]
```

```{r substr10.met.car.table.hisD.noalahis,results='asis',cache=FALSE}
printTable(subTbl10[order(pval,-1*original)], digits = 5,
           main = '$\\Delta hisD$ substrates',
           landscape = TRUE,
           #align = 'llllll',
           align = alEdgesOne,
           label = 'tab:SubsHisDnoalahis10')
```

#### Product table
```{r prepare.prd.table.hisD.noalahis}
prdTbl<-prnTbl[grepl('.+_u$',varname)&(original>0)]
prdTbl10<-prdTbl[(pval<=0.1)]
```

```{r prd10.met.car.table.hisD.noalahis,results='asis',cache=FALSE}
printTable(prdTbl10[order(pval,-1*original)], digits = 5,
           main = '$\\Delta hisD$ model products',
           landscape = TRUE,
           #align = 'llllll',
           align = alEdgesOne,
           label = 'tab:ProdsHisDnoalahis10')
```

#### Attranctant table
```{r prepare.attr.table.hisD.noalahis}
attrTbl<-prnTbl[grepl('.+_u$',varname)&(original<0)]
attrTbl10<-attrTbl[(pval<=0.1)]
```

```{r attr10.met.car.table.hisD.noalahis,results='asis',cache=FALSE}
printTable(attrTbl10[order(pval,original)], digits = 5,
           main = '$\\Delta hisD$ model attractants',
           landscape = TRUE,
           #align = 'llllll',
           align = alEdgesOne,
           label = 'tab:AttrsHisDnoalahis10')
```

#### Repellent table
```{r prepare.rep.table.hisD.noalahis}
repTbl<-prnTbl[grepl('.+_l$',varname)&(original<0)]
repTbl10<-repTbl[(pval<=0.1)]
```

```{r rep10.met.car.table.hisD.noalahis,results='asis',cache=FALSE}
printTable(repTbl10[order(pval,original)], digits = 5,
           main = '$\\Delta hisD$ model repellents',
           landscape = TRUE,
           #align = 'llllll',#
           align = alEdgesOne,
           label = 'tab:RepsHisDnoalahis10')
```


## Escherichia coli str. K12 substr. MG1655 ko trpB

The metabolic model of auxotrophic mutant $\Delta trpB$ was created by knocking out
Tryptophan synthase beta chain gene *trpB*, which suppose to control three 
reactions *TRPS1*, *TRPS2*, and *TRPS3*. However in the wild-type model 
*Escherichia_coli_str_K_12_substr_MG1655* genes seem to be duplicated: TRPS1 
is controlled by two genes *511145.12.1311_11.peg* and *511145.12.1310_10.peg*, 
while TRPS2 and TRPS3 are controlled by another two genes *511145.12.peg.1311* 
and *511145.12.peg.1310*. The second set of genes could be found on PATRIC, 
while the search for the first two genes gives no results. So the actual model was
created by removing three reactions *TRPS1*, *TRPS2*, and *TRPS3*.

### Read data

The following code loads simulation data:
```{r read.rds.trpB,eval=TRUE,cache=TRUE}
d<-"Escherichia_coli_str_K_12_substr_MG1655_ko_trpB"
df<-fread(file = file.path('data',d,'prcc_64.csv'))
```

```{r read.active.df16.trpB,echo=FALSE,cache=FALSE,eval=TRUE}
dt<-fread(file=file.path('data',d,'fba_sobol_isactive.csv'))
dtTrpB64<-dt
names(dtTrpB64)<-sub('^EX_','actEX_',names(dtTrpB64))
idx<-match(names(dt)[-c(1,ncol(dt))],df$varname)
```

```{r calc.active.df16.trpB,echo=FALSE,cache=TRUE,eval=TRUE}
df$PercActive16[idx]<-apply(as.data.frame(dt)[1:131072,-c(1,ncol(dt))],2,mean)
df$NActive16[idx]<-apply(as.data.frame(dt)[1:131072,-c(1,ncol(dt))],2,sum)
```
```{r calc.active.df32.trpB,echo=FALSE,cache=TRUE,eval=TRUE}
df$PercActive32[idx]<-apply(as.data.frame(dt)[1:262144,-c(1,ncol(dt))],2,mean)
df$NActive32[idx]<-apply(as.data.frame(dt)[1:262144,-c(1,ncol(dt))],2,sum)
```

```{r calc.active.df64.trpB,echo=FALSE,cache=TRUE,eval=TRUE}
df$PercActive64[idx]<-apply(as.data.frame(dt)[,-c(1,ncol(dt))],2,mean)
df$NActive64[idx]<-apply(as.data.frame(dt)[,-c(1,ncol(dt))],2,sum)
```


```{r filter.df.trpB}
dfAll<-rbind(dfAll,df,fill=TRUE )
dfw<-merge(dfw,df,by=mvar,all.x=TRUE,suffixes = c('.hisD2','.trpB'))
```
```{r hist.plot.trpB,fig.width=8,fig.height=8,fig.cap="\\label{fig:trpB_prcc_pval} Distribution of the significance of the exchange reaction boundaries PRCC values for the $\\Delta trpB$ model."}
qplot(df$pval,bins = 50)+
    xlab('p-value')+
    #ggtitle('Full dataset PRCC significance distribution for trpB')+
    theme_bw()
```

### Tryptophan flux plots

Similar to Supplementary Figures \ref{fig:wt_trp_bs}, \ref{fig:wt_trp_bs_f}, 
\ref{fig:wt_trp_bf}, and \ref{fig:wt_trp_bf_f}, which show influence of the 
tryptophan influx on the wild-type model, Supplementary Figures \ref{fig:trpB_trp_bs}, \ref{fig:trpB_trp_bs_f}, \ref{fig:trpB_trp_bf}, and \ref{fig:trpB_trp_bf_f} 
show the influence of the tryptophan influx on the $\Delta trpB$ model.

```{r prepare.trpB.plot.dataset}
fluxes_trpB<-fread(file=file.path('data',d,'mutant_related_fluxes.csv'))
re_names<-c("EX_trp_L(e)", "TRPAS2i", "TRPTA", "TRP_Ltex", "TRPt2rpp")
trpB<-fluxes_trpB[,match(paste0(re_names,'_fl'),names(fluxes_trpB)),with=FALSE]
pltDFtrp<-data.frame(Solution=fluxes_trpB$Solution,
                  influx_TrpL=-1*(trpB$`EX_trp_L(e)_fl`),
                  inBound_TrpL=1000*fluxes_trpB$`EX_trp_L(e)_l`,
                  outBound_TrpL=1000*fluxes_trpB$`EX_trp_L(e)_u`,
                  activityInflux=factor(dtTrpB64$`actEX_trp_L(e)_l`,levels = c(1,0),labels = c('active','inactive')),
                  activityOutflux=factor(dtTrpB64$`actEX_trp_L(e)_u`,levels = c(1,0),labels = c('active','inactive')))
```

```{r plot.trpB.bndvssol,fig.format='png',fig.dpi=600,fig.cap="\\label{fig:trpB_trp_bs} Influence of the tryptophan influx boundary on the biomass production in the $\\Delta trpB$ model. Sobol points where the boundary was active shown in red."}
qplot(inBound_TrpL,Solution,data = pltDFtrp,color=activityInflux,log='xy')+
    #ggtitle(bquote("Tryptophan influx boundary in "~Delta~"trpB mutant"))+
    theme_bw()
```

```{r plot.trpB.bndvssol_facet,fig.format='png',fig.dpi=600,fig.cap="\\label{fig:trpB_trp_bs_f} Influence of the tryptophan influx boundary on the biomass production in the $\\Delta trpB$ model. Sobol points where the boundary was active shown in red and placed on different panels."}
qplot(inBound_TrpL,Solution,data = pltDFtrp,color=activityInflux,log='xy',
      facets = activityInflux ~ .)+
    #ggtitle(bquote("Tryptophan influx boundary in "~Delta~"trpB mutant"))+
    theme_bw()
```

```{r plot.trpB.bndvsinflux,fig.format='png',fig.dpi=600,fig.cap="\\label{fig:trpB_trp_bf} Influence of the tryptophan influx boundary on the actual tryptophan influx value in the $\\Delta trpB$ model. Sobol points where the boundary was active shown in red."}
qplot(inBound_TrpL,influx_TrpL,data = pltDFtrp,color = activityInflux,log='xy')+
    #ggtitle(bquote("Tryptophan influx in "~Delta~"trpB mutant"))+
    theme_bw()
```

```{r plot.trpB.bndvsinflux.facet,fig.format='png',fig.dpi=600,fig.cap="\\label{fig:trpB_trp_bf_f} Influence of the tryptophan influx boundary on the actual tryptophan influx value in the $\\Delta trpB$ model. Sobol points where the boundary was active shown in red and placed on different panels."}
qplot(inBound_TrpL,influx_TrpL,data = pltDFtrp,color = activityInflux,log='xy',
      facets = activityInflux ~ .)+
    #ggtitle(bquote("Tryptophan influx in "~Delta~"trpB mutant"))+
    theme_bw()
```

### Interaction type table

We group all significantly sensitive exchange reaction boundary values and 
corresponding metabolites into four groups according to the type of boundary 
(upper or lower) and sign of the significance coefficient (PRCC). Number of
reactions in each group for $\Delta trpB$ is shown in the Supplementary Table \ref{tab:trpB_int_fq}.

Metabolites, which exchange reaction lower or upper boundary have a positive sensitivity coefficient, 
we will call substrates and products correspondingly. Substrates and products for the
$\Delta trpB$ model shown in Supplementary Tables \ref{tab:SubsTrpB10} and \ref{tab:ProdsTrpB10}.

Metabolites, which exchange reaction lower or upper boundary have a negative sensitivity coefficient, 
we will call repellents and attractants correspondingly. Repellents and attractants for the
$\Delta trpB$ model shown in Supplementary Tables \ref{tab:RepsTrpB10} and \ref{tab:AttrsTrpB10}.

```{r interaction.types.trpB}
prnTbl<-df[(card>1),c(1:3,6,31,32,34,36,8)]
prnTbl10<-prnTbl[(pval<=0.1)]
pander(table(sign(prnTbl10$original),sub('.+_(.+)$','\\1',prnTbl10$varname)),
       caption = '\\label{tab:trpB_int_fq} Sign/bound frequency table for $\\Delta trpB$ model')
```

#### Substrate table
```{r prepare.substr.table.trpB}
subTbl<-prnTbl[grepl('.+_l$',varname)&(original>0)]
subTbl10<-subTbl[(pval<=0.1)]
```

```{r substr10.met.car.table.trpB,results='asis',cache=FALSE}
printTable(subTbl10[order(pval,-1*original)], digits = 5,
           main = '$\\Delta trpB$ model substrates',
           landscape = TRUE,align = alEdgesOne,
           label = 'tab:SubsTrpB10')
```

#### Product table
```{r prepare.prd.table.trpB}
prdTbl<-prnTbl[grepl('.+_u$',varname)&(original>0)]
prdTbl10<-prdTbl[(pval<=0.1)]
```

```{r prd10.met.car.table.trpB,results='asis',cache=FALSE}
printTable(prdTbl10[order(pval,-1*original)], digits = 5,
           main = '$\\Delta trpB$ model products',
           landscape = TRUE,align = alEdgesOne,
           label = 'tab:ProdsTrpB10')
```

#### Attranctant table
```{r prepare.attr.table.trpB}
attrTbl<-prnTbl[grepl('.+_u$',varname)&(original<0)]
attrTbl10<-attrTbl[(pval<=0.1)]
```

```{r attr10.met.car.table.trpB,results='asis',cache=FALSE}
printTable(attrTbl10[order(pval,original)], digits = 5,
           main = '$\\Delta trpB$ model attractants',
           landscape = TRUE,align = alEdgesOne,
           label = 'tab:AttrsTrpB10')
```

#### Repellent table
```{r prepare.rep.table.trpB}
repTbl<-prnTbl[grepl('.+_l$',varname)&(original<0)]
repTbl10<-repTbl[(pval<=0.1)]
```

```{r rep10.met.car.table.trpB,results='asis',cache=FALSE}
printTable(repTbl10[order(pval,original)], digits = 5,
           main = '$\\Delta trpB$ model repellents',
           landscape = TRUE,align = alEdgesOne,
           label = 'tab:RepsTrpB10')
```

# Sensome Intersections
Supplementary Figures \ref{fig:upsetSens} and \ref{fig:vennSens} show how fluxes
that significantly influences biomass production distributed among different model
as UpSet plot [@Lex:2014UpSet; @Conway:2017UpSetR] and Venn diagram respectively.
It could be seen that majority of sensitive fluxes are common for all models,
fluxes, which are significantly influences biomass production not in all models,
listed in the Supplementary Table \ref{tab:upset_flx}. Most of metabolites listed
in the Supplementary Table \ref{tab:upset_flx} related to Drug metabolism, but
two things could be noticed: (i) the $\Delta hisD$ model is not sensitive to 
outflux of 2-Oxobutanoate unlike all other three models, and (ii) 
the $\Delta trpB$ is sensitive to influx of D-glucito and outflux of the Tiamin
apart from expected influx of L-tryptophan.


```{r upset.plot, fig.width=8,fig.height=8,fig.cap="\\label{fig:upsetSens} UpSet representation of intersection of sets of fluxes, for which PRCC different from zero at the significance level of 10% in different models: wt -- the wild-type model, hisD -- the $\\Delta hisD$ model, hisD2 -- the $\\Delta hisD$ model with blocked AlaHis assimilation, trpB -- the $\\Delta trpB$ model."}
upsetinput<-list(wt=dfw$varname[dfw$pval.wt<0.10],
                 hisD=dfw$varname[dfw$pval.hisD<0.1],
                 hisD2=dfw$varname[dfw$pval.hisD2<0.1],
                 trpB=dfw$varname[dfw$pval.trpB<0.1])
upset(fromList(upsetinput))
```

```{r venn.plot, fig.width=8,fig.height=8,fig.cap="\\label{fig:vennSens} Venn plot of sets of fluxes, for which PRCC different from zero at the significance level of 10% in different models: wt -- the wild-type model, hisD -- the $\\Delta hisD$ model, hisD2 -- the $\\Delta hisD$ model with blocked AlaHis assimilation, trpB -- the $\\Delta trpB$ model."}
vn<-gplots::venn(upsetinput)
```

```{r uncommon.sensitive.fluxes}
ints<-attr(vn,'intersections')
df1<-dfw[match(ints$`wt:hisD:trpB`,df$varname),c("MetID","varname","Name","ExEquation")]
df1$Group<-'wt:hisD:trpB'
df2<-dfw[match(ints$`wt:hisD2:trpB`,df$varname),c("MetID","varname","Name","ExEquation")]
df2$Group<-'wt:hisD2:trpB'
df3<-dfw[match(ints$trpB,df$varname),c("MetID","varname","Name","ExEquation")]
df3$Group<-'trpB'
df4<-dfw[match(ints$hisD2,df$varname),c("MetID","varname","Name","ExEquation")]
df4$Group<-'hisD2'
intsDF<-rbind(df1,df2,df3,df4)[,c('Group',"MetID","varname","Name","ExEquation")]
pander(intsDF,
       caption=paste('\\label{tab:upset_flx} Sets of fluxes, for which PRCC ',
                     'different from zero at the significance level of 10% ',
                     'not in all models. Model names: wt -- the wild-type model, ',
                     'hisD -- the $\\Delta hisD$ model, hisD2 -- the $\\Delta hisD$ ',
                     'model with blocked AlaHis assimilation, trpB -- the ',
                     '$\\Delta trpB$ model.'))
```


# Bidirectional metabolites

```{r filter.df}
df10<-dfAll[pval<0.1]
df<-dfAll
```

```{r bidir.met.p10}
mdf10<-df10[,.(N=length(unique(varname))),by=.(Strain,MetID,Name)][N>1]
fmdf10<-merge(mdf10,df10,by=c('Strain','MetID','Name'))
dlgpval10<-fmdf10[,.(dlg=mean(abs(diff(log10(pval)))),
                     sgn=prod(sign(original))),
                  by=.(Strain,MetID,Name)]
fmdf10<-merge(fmdf10,dlgpval10,by=c('Strain','MetID','Name'))
metCar<-fmdf10[,.(N=length(unique(Strain)),
                  Nc=length(unique(Strain[which(card==1)])),
                  Nr=length(unique(Strain[which(range<=1e-5)])),
                  Nlgdif=length(unique(Strain[which(dlg<1)])),
                  Nsign=length(unique(Strain[which(sgn>0)]))),
               by=.(MetID,Name)]
```

```{r bidir.met.car.table.p10,results='asis',cache=FALSE}
printTable(metCar[order(metCar$N,decreasing = TRUE),],
           main = 'Frequency of bidirectional metabolites',
           landscape = FALSE,align = alMetCarF,
           label = 'tab:bidir_met_p10')

```

### Glycerol 3-phosphate
```{r glyc3p.activity.wt}
t<-as.data.frame(table(dtWT64$`actEX_glyc3p(e)_l`,dtWT64$`actEX_glyc3p(e)_u`))
names(t)<-c('Lower bound active','Upper bound active','Count')
levels(t$`Lower bound active`)<-c('No','Yes')
levels(t$`Upper bound active`)<-c('No','Yes')
t$Percent<-t$Count*100/nrow(dtWT64)
pander(t,
       caption=paste('\\label{tab:glyc3p_cnt} Fractions of Sobol points groupped according to',
                     'Glycerol 3-phosphate lower and upper bound activity'))
```

### Succinate
```{r succ.activity.wt}
t<-as.data.frame(table(dtWT64$`actEX_succ(e)_l`,dtWT64$`actEX_succ(e)_u`))
names(t)<-c('Lower bound active','Upper bound active','Count')
levels(t$`Lower bound active`)<-c('No','Yes')
levels(t$`Upper bound active`)<-c('No','Yes')
t$Percent<-t$Count*100/nrow(dtWT64)
pander(t,
       caption=paste('\\label{tab:succ_cnt} Fractions of Sobol points groupped according to',
                     'Succinate lower and upper bound activity'))
```

### Trimethylamine N-oxide
```{r tmao.activity.wt}
t<-as.data.frame(table(dtWT64$`actEX_tmao(e)_l`,dtWT64$`actEX_tmao(e)_u`))
names(t)<-c('Lower bound active','Upper bound active','Count')
levels(t$`Lower bound active`)<-c('No','Yes')
levels(t$`Upper bound active`)<-c('No','Yes')
t$Percent<-t$Count*100/nrow(dtWT64)
pander(t,
       caption=paste('\\label{tab:succ_cnt} Fractions of Sobol points groupped according to',
                     'Trimethylamine N-oxide lower and upper bound activity'))
```

## Interaction type table for all three models
```{r interaction.types}
prnTbl<-dfw[(card.wt>1)|(card.hisD>1)|(card.trpB>1),
            c(1:2,
              grep('original',names(dfw)),
              grep('pval',names(dfw)),
              grep('card',names(dfw)),
              grep('Name',names(dfw))),with=FALSE]
prnTbl10<-prnTbl[(pval.wt<=0.1)]
writeExcel<-!file.exists(excelFileName)
pander(table(sign(prnTbl10$original.wt),sub('.+_(.+)$','\\1',prnTbl10$varname)),
       caption = '\\label{tab:all_int_fq} Aggregated sign/bound frequency table for all models')
```

```{r write.excel,eval=writeExcel}
subTbl<-prnTbl[grepl('.+_l$',varname)&((original.wt>0)|(original.hisD>0)|(original.trpB>0))]
prdTbl<-prnTbl[grepl('.+_u$',varname)&((original.wt>0)|(original.hisD>0)|(original.trpB>0))]
attrTbl<-prnTbl[grepl('.+_u$',varname)&((original.wt<0)|(original.hisD<0)|(original.trpB<0))]
repTbl<-prnTbl[grepl('.+_l$',varname)&((original.wt<0)|(original.hisD<0)|(original.trpB<0))]
l<-list(Substrates=subTbl,Products=prdTbl,Attractants=attrTbl,Repellents=repTbl)
write.xlsx(l, file = excelFileName)
```

\pagebreak

# References

<div id="refs"></div>

\pagebreak

# Appendix {.tabset}
## Functions
```{r functions, eval=FALSE, include=TRUE}
```

## Setup R
```{r setup, eval=FALSE}
```

## Versions

### Session Info
```{r sessionInfo, echo=FALSE, results='asis', class='text', warning=FALSE}
si<-devtools::session_info()
cat('Platform\n\n')
pander::pander(si$platform)
cat('Packages\n\n')
knitr::kable(as.data.frame(si$packages)[,c('ondiskversion','loadedversion','date','source')],align = c('l','l'))
```

## List of figures

\listoffigures

## List of tables

\listoftables
